public/app/features/explore/Graph.tsx

```diff
@@ -143,52 +174,64 @@ export class Graph extends PureComponent<GraphProps, GraphState> {
     );
   };
 
-  draw() {
-    const { range, size, userOptions = {} } = this.props;
+  onToggleSeries = (series: TimeSeries, exclusive: boolean) => {
+    this.setState(
+      (state, props) => {
+        const { data } = props;
+        const { hiddenSeries } = state;
+        const hidden = hiddenSeries.has(series);
+        const oneSeriesVisible = hiddenSeries.size === data.length - 1;
+        if (exclusive) {
+          return {
+            hiddenSeries: !hidden && oneSeriesVisible ? new Set() : new Set(data.filter(d => d !== series)),
+          };
+        }
+        const nextHiddenSeries = new Set(hiddenSeries);
+        if (nextHiddenSeries.has(series)) {
+          nextHiddenSeries.delete(series);
+        } else {
+          nextHiddenSeries.add(series);
+        }
+        return {
+          hiddenSeries: nextHiddenSeries,
+        };
+      },
+      () => this.draw(false)
```

Comment:
wont set state cause a componentDidUpdate & draw already?

Reply:
In the current state, `draw` would not be called because componentDidUpdate only calls it when certain things changed, and `hiddenSeries` it not among them.

It's a valid point though, to gate the execution of `draw` through componentDidUpdate like it is for other props/state items. I'd add a check for `hiddenSeries` there and get rid of the `updateDynamicOptions` logic.

Reply:
Ah, thanks for the advice on `componentDidUpdate`—my rookie React skills definitely showing here.

My original intention of having `updateDynamicOptions` is due to the `range` prop that's fed into the `Graph` component. It seems to constantly update itself. This meant toggling series also undesirably advanced the time range for the x-axis.

Any advice here? It's like I need to cache the last `dynamicOptions` that came about from running a query.